File: kitchen-ai/java/src/main/java/ai/kitchen/MenuPlanner.java
package ai.kitchen;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class MenuPlanner {
    private static final List<String> DAY_NAMES = List.of(
            "Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag", "Sonntag"
    );

    private static final Map<String, List<String>> SUBSTITUTIONS = Map.of(
            "Babyspinat", List.of("TK-Spinat", "Rucola"),
            "Tahini", List.of("Erdnussmus", "Cashewmus"),
            "Parmesan", List.of("Hefeflocken", "Pekorino"),
            "Pinienkerne", List.of("Sonnenblumenkerne", "Mandeln"),
            "Tofu", List.of("Tempeh", "Kichererbsen"),
            "Weißfisch", List.of("Lachs", "Kabeljau"),
            "Basilikum", List.of("Petersilie", "TK-Kräuter"),
            "Currypaste", List.of("Currypulver", "Garam Masala")
    );

    private record CoverageResult(double coverage, List<MissingItem> missing) {}

    public static void main(String[] args) throws IOException {
        Path recipesPath = Paths.get("src/main/resources/recipes.json");
        Path ingredientsPath = Paths.get("src/main/resources/available_ingredients.json");
        int servings = 2;
        int days = 7;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--recipes" -> recipesPath = Paths.get(args[++i]);
                case "--ingredients" -> ingredientsPath = Paths.get(args[++i]);
                case "--servings" -> servings = Integer.parseInt(args[++i]);
                case "--days" -> days = Integer.parseInt(args[++i]);
                default -> {
                }
            }
        }

        ObjectMapper mapper = new ObjectMapper();
        List<Recipe> recipes = loadRecipes(mapper, recipesPath);
        Map<String, InventoryItem> available = loadInventory(mapper, ingredientsPath);

        List<PlanItem> plan = planWeek(recipes, available, days, servings);
        Map<String, ShoppingEntry> shopping = buildShoppingList(plan);

        System.out.println("Menüplan\n");
        System.out.println(formatPlan(plan));
        System.out.println();
        System.out.println(formatShopping(shopping));
    }

    private static List<Recipe> loadRecipes(ObjectMapper mapper, Path path) throws IOException {
        return mapper.readValue(path.toFile(), new TypeReference<>() {});
    }

    private static Map<String, InventoryItem> loadInventory(ObjectMapper mapper, Path path) throws IOException {
        return mapper.readValue(path.toFile(), new TypeReference<>() {});
    }

    private static CoverageResult computeCoverage(Recipe recipe, Map<String, InventoryItem> available, int targetServings) {
        List<Ingredient> required = recipe.getIngredients().stream()
                .filter(i -> !i.isOptional())
                .toList();

        List<MissingItem> missing = new ArrayList<>();
        int covered = 0;

        for (Ingredient ingredient : required) {
            double requiredQty = scaleQuantity(ingredient.getQuantity(), recipe.getServings(), targetServings);
            InventoryItem entry = available.get(ingredient.getName());
            if (entry != null && entry.getQuantity() >= requiredQty) {
                covered += 1;
            } else {
                double availableQty = entry == null ? 0 : entry.getQuantity();
                missing.add(new MissingItem(ingredient.getName(), requiredQty, ingredient.getUnit(), availableQty));
            }
        }
        double coverage = required.isEmpty() ? 1.0 : (double) covered / required.size();
        return new CoverageResult(coverage, missing);
    }

    private static double scaleQuantity(double quantity, int recipeServings, int targetServings) {
        double factor = (double) targetServings / recipeServings;
        return Math.round(quantity * factor * 100.0) / 100.0;
    }

    private static List<PlanItem> planWeek(List<Recipe> recipes, Map<String, InventoryItem> available, int days, int targetServings) {
        List<Map<String, Object>> scored = new ArrayList<>();
        for (Recipe recipe : recipes) {
            CoverageResult result = computeCoverage(recipe, available, targetServings);
            Map<String, Object> entry = new HashMap<>();
            entry.put("recipe", recipe);
            entry.put("coverage", result.coverage());
            entry.put("missing", result.missing());
            scored.add(entry);
        }

        scored.sort(Comparator
                .comparing((Map<String, Object> m) -> (Double) m.get("coverage"))
                .thenComparing(m -> -((List<?>) m.get("missing")).size())
                .reversed());

        List<Map<String, Object>> candidates = scored.stream()
                .filter(m -> (Double) m.get("coverage") >= 0.8)
                .toList();
        if (candidates.isEmpty()) {
            candidates = scored.stream().limit(3).toList();
        }

        List<PlanItem> plan = new ArrayList<>();
        for (int i = 0; i < days; i++) {
            Map<String, Object> entry = candidates.get(i % candidates.size());
            Recipe recipe = (Recipe) entry.get("recipe");
            double coverage = Math.round(((Double) entry.get("coverage")) * 100.0) / 100.0;
            @SuppressWarnings("unchecked")
            List<MissingItem> missing = (List<MissingItem>) entry.get("missing");
            plan.add(new PlanItem(DAY_NAMES.get(i % DAY_NAMES.size()), recipe, coverage, missing));
        }
        return plan;
    }

    private static Map<String, ShoppingEntry> buildShoppingList(List<PlanItem> plan) {
        Map<String, ShoppingEntry> shopping = new HashMap<>();
        for (PlanItem item : plan) {
            for (MissingItem missing : item.getMissing()) {
                ShoppingEntry entry = shopping.getOrDefault(missing.getName(), new ShoppingEntry(missing.getUnit(), 0));
                entry.addQuantity(missing.getRequired());
                shopping.put(missing.getName(), entry);
            }
        }
        return shopping;
    }

    private static List<String> suggestSubstitutions(List<MissingItem> missing) {
        List<String> suggestions = new ArrayList<>();
        for (MissingItem item : missing) {
            List<String> subs = SUBSTITUTIONS.get(item.getName());
            if (subs != null && !subs.isEmpty()) {
                suggestions.add(item.getName() + ": ersetze durch " + String.join(", ", subs));
            }
        }
        return suggestions;
    }

    private static String formatPlan(List<PlanItem> plan) {
        StringBuilder builder = new StringBuilder();
        for (PlanItem item : plan) {
            builder.append(String.format("%s: %s (Abdeckung: %d%%)%n",
                    item.getDay(), item.getRecipe().getTitle(), (int) (item.getCoverage() * 100)));
            if (item.getMissing().isEmpty()) {
                builder.append("  Alle Zutaten vorhanden.\n");
            } else {
                builder.append("  Fehlende Zutaten:\n");
                for (MissingItem missing : item.getMissing()) {
                    builder.append(String.format(
                            "    - %s: benötigt %.2f %s, verfügbar %.2f%n",
                            missing.getName(), missing.getRequired(), missing.getUnit(), missing.getAvailable()
                    ));
                }
                List<String> substitutions = suggestSubstitutions(item.getMissing());
                if (!substitutions.isEmpty()) {
                    builder.append("  Vorschläge für Substitution:\n");
                    substitutions.forEach(sub -> builder.append("    - ").append(sub).append("\n"));
                }
            }
        }
        return builder.toString();
    }

    private static String formatShopping(Map<String, ShoppingEntry> shopping) {
        if (shopping.isEmpty()) {
            return "Keine zusätzlichen Einkäufe nötig.";
        }
        return "Einkaufsliste (kumuliert):\n" + shopping.entrySet().stream()
                .map(entry -> String.format("- %s: %.2f %s", entry.getKey(), entry.getValue().quantity(), entry.getValue().unit()))
                .collect(Collectors.joining("\n"));
    }

    private static class ShoppingEntry {
        private final String unit;
        private double quantity;

        ShoppingEntry(String unit, double quantity) {
            this.unit = unit;
            this.quantity = quantity;
        }

        void addQuantity(double add) {
            this.quantity = Math.round((this.quantity + add) * 100.0) / 100.0;
        }

        public String unit() {
            return unit;
        }

        public double quantity() {
            return quantity;
        }
    }
}

File: kitchen-ai/python/menu_planner.py
import argparse
import json
import math
from pathlib import Path
from typing import Dict, List, Tuple

Ingredient = Dict[str, object]
Recipe = Dict[str, object]

DAY_NAMES = ["Montag", "Dienstag", "Mittwoch", "Donnerstag", "Freitag", "Samstag", "Sonntag"]
SUBSTITUTIONS = {
    "Babyspinat": ["TK-Spinat", "Rucola"],
    "Tahini": ["Erdnussmus", "Cashewmus"],
    "Parmesan": ["Hefeflocken", "Pekorino"],
    "Pinienkerne": ["Sonnenblumenkerne", "Mandeln"],
    "Tofu": ["Tempeh", "Kichererbsen"],
    "Weißfisch": ["Lachs", "Kabeljau"],
    "Basilikum": ["Petersilie", "TK-Kräuter"],
    "Currypaste": ["Currypulver", "Garam Masala"],
}


def load_json(path: Path):
    with path.open(encoding="utf-8") as handle:
        return json.load(handle)


def scale_quantity(quantity: float, recipe_servings: int, target_servings: int) -> float:
    factor = target_servings / recipe_servings
    return round(quantity * factor, 2)


def compute_coverage(
    recipe: Recipe, available: Dict[str, Dict[str, object]], target_servings: int
) -> Tuple[float, List[Dict[str, object]]]:
    required = [i for i in recipe["ingredients"] if not i.get("optional")]
    missing: List[Dict[str, object]] = []
    covered = 0

    for ingredient in required:
        name = ingredient["name"]
        required_qty = scale_quantity(float(ingredient["quantity"]), recipe["servings"], target_servings)
        available_entry = available.get(name)
        if available_entry and float(available_entry.get("quantity", 0)) >= required_qty:
            covered += 1
        else:
            missing.append(
                {
                    "name": name,
                    "required": required_qty,
                    "unit": ingredient.get("unit", ""),
                    "available": available_entry.get("quantity") if available_entry else 0,
                }
            )
    coverage = covered / len(required) if required else 1.0
    return coverage, missing


def plan_week(recipes: List[Recipe], available: Dict[str, Dict[str, object]], days: int, target_servings: int) -> List[Dict[str, object]]:
    scored = []
    for recipe in recipes:
        coverage, missing = compute_coverage(recipe, available, target_servings)
        scored.append({"recipe": recipe, "coverage": coverage, "missing": missing})

    scored.sort(key=lambda x: (x["coverage"], -len(x["missing"])), reverse=True)

    candidates = [entry for entry in scored if entry["coverage"] >= 0.8]
    if not candidates:
        candidates = scored[:3]

    plan = []
    for day_index in range(days):
        entry = candidates[day_index % len(candidates)]
        plan.append({
            "day": DAY_NAMES[day_index % len(DAY_NAMES)],
            "recipe": entry["recipe"],
            "coverage": round(entry["coverage"], 2),
            "missing": entry["missing"],
        })
    return plan


def build_shopping_list(plan: List[Dict[str, object]]) -> Dict[str, Dict[str, float]]:
    shopping: Dict[str, Dict[str, float]] = {}
    for item in plan:
        for missing in item["missing"]:
            name = missing["name"]
            if name not in shopping:
                shopping[name] = {"quantity": 0, "unit": missing.get("unit", "")}
            shopping[name]["quantity"] += missing.get("required", 0)
    return shopping


def suggest_substitutions(missing: List[Dict[str, object]]) -> List[str]:
    suggestions = []
    for item in missing:
        subs = SUBSTITUTIONS.get(item["name"])
        if subs:
            suggestions.append(f"{item['name']}: ersetze durch {', '.join(subs)}")
    return suggestions


def format_plan(plan: List[Dict[str, object]]):
    lines = []
    for item in plan:
        recipe = item["recipe"]
        lines.append(f"{item['day']}: {recipe['title']} (Abdeckung: {int(item['coverage'] * 100)}%)")
        if item["missing"]:
            lines.append("  Fehlende Zutaten:")
            for missing in item["missing"]:
                lines.append(
                    f"    - {missing['name']}: benötigt {missing['required']} {missing.get('unit', '')}, verfügbar {missing.get('available', 0)}"
                )
            subs = suggest_substitutions(item["missing"])
            if subs:
                lines.append("  Vorschläge für Substitution:")
                for sub in subs:
                    lines.append(f"    - {sub}")
        else:
            lines.append("  Alle Zutaten vorhanden.")
    return "\n".join(lines)


def format_shopping(shopping: Dict[str, Dict[str, float]]):
    if not shopping:
        return "Keine zusätzlichen Einkäufe nötig."
    lines = ["Einkaufsliste (kumuliert):"]
    for name, detail in shopping.items():
        qty = round(detail["quantity"], 2)
        unit = detail.get("unit", "")
        lines.append(f"- {name}: {qty} {unit}")
    return "\n".join(lines)


def main():
    parser = argparse.ArgumentParser(description="Erzeuge einen Wochen-Menüplan aus Rezept-JSON und verfügbaren Zutaten.")
    parser.add_argument("--recipes", default="recipes.json", type=Path, help="Pfad zur Rezept-JSON")
    parser.add_argument("--ingredients", default="available_ingredients.json", type=Path, help="Pfad zur Zutaten-JSON")
    parser.add_argument("--servings", default=2, type=int, help="Zielportionen pro Rezept")
    parser.add_argument("--days", default=7, type=int, help="Anzahl der Tage für den Plan")
    args = parser.parse_args()

    recipes = load_json(args.recipes)
    available = load_json(args.ingredients)

    plan = plan_week(recipes, available, args.days, args.servings)
    shopping_list = build_shopping_list(plan)

    print("Menüplan:\n")
    print(format_plan(plan))
    print("\n" + format_shopping(shopping_list))


if __name__ == "__main__":
    main()
